Wikipedia/Code/WMFCVLInfo.h

```diff
@@ -0,0 +1,43 @@
+#import <Foundation/Foundation.h>
+
+@class WMFCVLColumn;
+@class WMFCVLSection;
+@class WMFCVLAttributes;
+@class WMFCVLInvalidationContext;
+@class WMFColumnarCollectionViewLayout;
+@class WMFCVLMetrics;
+@protocol WMFColumnarCollectionViewLayoutDelegate;
+
+/*!
+ @class        WMFCVLInfo
+ @abstract     A WMFCVLInfo is a snapshot of a WMFColumnarCollecitonViewLayout and the layout attributes of the items and supplementary views. It handles the layout and organization into colums based on the WMFCVLMetrics provided to it.
+ @discussion   ...
+ */
+@interface WMFCVLInfo : NSObject <NSCopying>
+
+@property (nonatomic, strong, nonnull, readonly) NSArray <WMFCVLColumn *> *columns;
+@property (nonatomic, strong, nonnull, readonly) NSArray <WMFCVLSection *> *sections;
+
+@property (nonatomic) CGSize boundsSize;
+@property (nonatomic) CGSize contentSize;
+
+/*!
+ Initialize a new WMFCVLInfo with the given metrics.
+ @param metrics     The metrics for the WMFCVLInfo
+ @return A WMFCVLInfo with the given metrics
+ */
+- (nonnull instancetype)initWithMetrics:(nonnull WMFCVLMetrics *)metrics NS_DESIGNATED_INITIALIZER;
+
+- (void)enumerateSectionsWithBlock:(nonnull void(^)(WMFCVLSection * _Nonnull section, NSUInteger idx, BOOL * _Nonnull stop))block;
+- (void)enumerateColumnsWithBlock:(nonnull void(^)(WMFCVLColumn * _Nonnull column, NSUInteger idx, BOOL * _Nonnull stop))block;
+
+- (nullable WMFCVLAttributes *)layoutAttributesForItemAtIndexPath:(nonnull NSIndexPath *)indexPath;
+- (nullable WMFCVLAttributes *)layoutAttributesForSupplementaryViewOfKind:(nonnull NSString *)elementKind atIndexPath:(nonnull NSIndexPath *)indexPath;
+
+- (void)layoutForBoundsSize:(CGSize)size withDelegate:(nullable id <WMFColumnarCollectionViewLayoutDelegate>)delegate collectionView:(nullable UICollectionView *)collectionView;
```

Comment:
I don't think you could because this method iterates over the delegate and calls `estimatedHeightForHeaderInSection:` for each section. Also that would use make a long method. I think having access to the delegate is fine.

It's ok to pass them in every time, I just wasn't sure why the delegate was being passed on each call while the metrics was being retained strongly on init.  I guess if the metrics change you need to create a new instance of the info class, but if the delegate or collection view change that isn't necessary? Either way, either stateless or immutable is best, so this is ok to me.